13985: yaml validation doc (#14049)
CDCgov/prime-reportstream: prime-router/docs/design/proposals/0026-yaml-validation/0026-yaml-validation.md

# `YAML` Validation Proposal

## Background

Our application uses `YAML` in a variety of contexts for configuration. It would speed up development and reduce runtime 
errors if we were able to validate the `YAML` structure and value types. We want to catch these issues at compile-time during
development.

## Assumptions

* We will want to continue using `YAML` as our primary configuration language
* We will want our `YAML` validated every time we either read or update a configuration

## Possible Implementations

> ### Use Apple's `Pkl` configuration library ([link](https://pkl-lang.org/index.html))

Reviewer: there are other possibilities

Author: None of this is set in stone. I can be convinced if theres a better way!

we need a solution that works for the codebase we have, not the one we want.

### Write our own validation code with Jackson Serialization

#### Approach

* Create a data class structure that matches our configuration (starting with organizations.yml)
* Serialize it with a strict Jackson `YAML` mapper that throws errors on missing fields or extra fields
* Run custom validation code on it to ensure configuration values are acceptable

#### Pros

* This would be the most flexible approach for us as we could write validation code very specific to Report Stream that could not be captured otherwise
  * Example: ensure filter definitions are appropriately formatted

#### Cons

* None of the work will be outsourced to existing libraries and would have to all be done in-house
* The validations would be spread across multiple files in various places and could be hard to find
* The configuration data classes would have to change every time we made an update

#### Decision

This idea would work but may require a lot of custom work. This would be a good choice if we want very in-depth validation.

> ### Write JSON Schema files and run validations against it

Reviewer 2: This section should also mention Konform and how we plan to use it and how its different from JSON schema files?

Author: will do.

#### Pros

* Off the shelf standardized solution
* All validations should be contained in one spot for easy locating

#### Cons

* Lose out on extra validations for values beyond types

#### Decision

> This would be my preferred solution as it has the least upfront work with the greatest return. It does not

Reviewer 2: ```suggestion
This is presently the preferred solution as it has the least upfront work with the greatest return. It does not
```
